### ISSUE TITLE
Remote Flutter Widgets (rfw) package feedback thread

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#90218]
This issue is intended to collect feedback regarding the [rfw package](https://pub.dev/packages/rfw).

If you have found a bug with RFW, please [file a new issue](https://github.com/flutter/flutter/issues/new?assignees=Hixie&labels=p:%20rfw,package,P4). This issue is specifically for discussing the package's general applicability to your use cases, to determine if this is a useful direction for us to follow, and to report experiences you may have with writing applications with this package. Thanks in advance.

**Comment by shuoch-g** [comment]
Hi, Ian,

Very thrilled to see flutter team's effort on the dynamic UI ! We are evaluating on how to leverage rfw in our project.

First of all, may I ask about the compatibility policy we will have in rfw ?

In our project we will send down configuration data from server (`DynamicMap` analogue in rfw) and our runtime evaluator will translate them to widget constructor calls. As the evaluator code (decoder + `WidgetBuilder` analogue in rfw) is released with our app's apk, we want the server returned configuration to be always compatible with the evaluator that is released in a client version.

This is how we solve this: For each client release, we will snapshot the "capabilities" of the evaluator by the branch cut. The capabilities include

1. which set of widgets are supported, and their names.
2. which set of arguments are supported for each widget, and their name/type.

these snapshot capabilities data will be kept in our database, and keyed by the app release version.

Before the configuration data is sent to client with version A, we will retrieve the capabilities data for version A, and go through all configuration data and validate that all configuration key/value are supported in the capabilities. Our feature is authored in a single place, but when releasing to different client versions we will do tree-shaking or branching and release different configuration data targeting different client apk versions.

Could rfw package provide an interface that returns a structured data object that encodes all widgets / arguments its `WidgetBuilder`s can support ? If it makes sense to you, happy to contribute as well.

**Comment by Hixie** [comment]
Right now there's no compatibility policy but there could be one. I think it makes sense to have a policy that we never break things.

I don't know how easy it would be to accurately return what the supported API is. You can get the list of widgets easily enough by querying the map of constructors, but getting the arguments is non-trivial.

**Comment by ljmatan** [comment]
@Hixie 
My main problem with this package is the fact it does not resolve the issue from which it stemmed all while the effort spent on the package could've been utilised on much more important issues (e.g., web "app" size, web scroll, iOS everything (it's a disaster), not being able to conditionally import packages, etc.). 

In other words, package accomplishes nothing while there are many other glaring issues being disregarded while work is done on this abomination.

**Comment by Hixie** [comment]
@ljmatan Well the good news is that it didn't take much effort to make this package so it didn't take away from those other efforts. :-) That said, your comment really isn't constructive and doesn't really make anyone feel welcome, so please think of other ways to phrase that kind of feedback in the future. You may wish to review our code of conduct, which requires everyone to be actively welcoming. Feedback along those lines can be given in a much more friendly manner.

**Comment by ljmatan** [comment]
@Hixie 
Mr Hixie, 

You're actually absolutely correct on my conduct, and I do apologise. 

However, seeing as it's another half-measure taken after another half-measure which came after another half-measure, and seeing as you are a figure that's relevant within the Flutter community, I feel it's important for you to understand that the progress on the framework itself is severly lacking as very large issues are not handled in a timmely manner, all the while you're working on and promoting something which does not resolve aforementioned issues.

**Comment by fzyzcjy** [comment]
Related: https://github.com/fzyzcjy/flutter_rust_bridge/issues/296 Cooperate flutter_rust_bridge with package:rfw(Remote Flutter Widgets), for example, downloading the rust compiled library at runtime and binding with that.

**Comment by fzyzcjy** [comment]
@Hixie As suggested by you a few months ago, we are interested in cooperating rfw with flutter_rust_bridge. Could you please provide some suggestions on the details?

Related: https://github.com/fzyzcjy/flutter_rust_bridge/issues/296

**Comment by Hixie** [comment]
(responded in the other issue — thanks!)

**Comment by mhoad** [comment]
@Hixie I only just came across this package yesterday and haven’t had a chance to kick the tires on it yet but I’m incredibly optimistic about this direction especially when it comes to app delivery and I am curious to hear your thoughts about what kinds of ways this could be extended and built upon in the future.

One of the things I find hardest with Flutter web currently is just how much code I have to ship to get that first frame on the screen and the lack of modularity compared to say how I might deliver a regular HTML / JS / CSS based web app.

It also gives me more control it seems over what kind of experience I want to render for what kind of user which I appreciate because again I feel like currently I’m in a position where I need to ship all possible set of experiences not only when the user might only want a tiny fraction of them but I have to do all of this before I even render a frame which is upsetting for everyone.

I would love to see what a Flutter application server could look like.

**Comment by Hixie** [comment]
This technique doesn't reduce the amount of code you have to send down, really, since you still have to send the whole Flutter engine and the whole framework, but certainly something using this or another solution along this lines could give you more flexibility in terms of what the application does.

**Comment by dengzq** [comment]
Hi !  Thanks for doing this for flutter dynamic ui !  Here i set a config.json file to config all the widgets that we support.

**Comment by hmbenhaim** [comment]
Hi thanks for the package, I always wanted dynamic widget like you have in html,
Just wanted to ask will you add like regular imports for consistency, like ```import "package:flutter/material.dart"``` it will be a lot easier to manage the code this way,

**Comment by Hixie** [comment]
@hmbenhaim do you mean from an .rfwtxt file?

**Comment by 0xchase** [comment]
Hey, I just wanted to add that I've found this package useful in designing a (sort of) interactive UI designer for my app. I hope it continues to be developed and maintained.

**Comment by isAlmogK** [comment]
Hi, I have a couple of questions and hope someone from the team can help; in general, this could be very useful, especially if you need to load custom widgets for enterprise customers, at least in our use case. 

My first question is whether there are plans to support this package, as I see there has not been a big use case. The second question would it be possible to load the following into a widget build method from, let's say Firestore 

```
MultiSelect(
                onSaved: (val) {
                  setState(() {
                    for (final element in commandsType) {
                      final contain =
                          commandTypeName.where((item) => item == element.name);
                      if (contain.isEmpty) {
                        commandTypeName.add(element.name as String);
                      }
                    }
                    dynamicTrainingType.commands =
                        commandTypeName.cast<String>();
                  });
                },
                buttonText: 'Select Commands',
                items: _commandItems,
                initialVue: commandsType,
                onConfirm: (dynamic vr) {
                  setState(() {
                    dynamicTrainingType.commands?.clear();
                    commandsType = vr as List<dynamic>;
                  });
                },
              ),
```

**Comment by Hixie** [comment]
@isAlmogK Can you elaborate on what you mean by "support" exactly?

For the second question, I'm not sure I understand exactly what you are asking; in general package:rfw does not allow imperative code to be sent from the server, only declarative widgets. You could send code as JavaScript or Wasm or some such though. If you have a clearer description of your use case I could try to advise in more detail.

**Comment by Hixie** [comment]
Can you get the wasm package to work in isolation?

**Comment by dkbast** [comment]
-> looks like the hello_world example is running just fine

```
wasmtest ➤ dart run                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       git:main*
Building package executable... 
Built wasmtest:wasmtest.
1522756 == 1522756
```

had to reinstall rust, but still no luck with rfw, did initialise wasm with `flutter pub run` as well, so this is not the issue.

**Comment by Hixie** [comment]
I have no idea... does it work on a Linux box? On this Linux machine, modulo some other issue I'm having with LIBGL, if I just check out the rfw repo and go into the example/wasm directory and run `flutter pub run wasm:setup; flutter run -d linux` it seems to work.

**Comment by dkbast** [comment]
Just hopped on my old Macbook with Intel processor and got the same error - the fact that the debug entitlements for network access are missing from `macos/Runner/DebugProfile.entitlements` leads me to believe that this has not been tested for macos? Because the example as is, is not runnable on a mac without adding this:

```
<key>com.apple.security.network.client</key>
<true/>
```

Currently I don't have access to another machine, but if it works on linux for you, thats good - that probably narrows this issue down to being mac specific (but not M1 mac specific)

**Comment by Hixie** [comment]
Aah, yes, it is probably that exact problem, indeed.

**Comment by dkbast** [comment]
I just created a small script to checkout and run the wasm example on a mac machine, to make this easier to test and reproduce: https://gist.github.com/dkbast/6af0c48eda59817b525798743f737918 

```Bash
# A shell script which sets everything up to run the rfw wasm example
# Setup Rust:
brew install rustup
rustup-init
source "$HOME/.cargo/env"
# clone the project
git clone https://github.com/flutter/packages.git

cd packages/packages/rfw

cd example/wasm

# Add internet access to the debug entitlements:
echo '<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>com.apple.security.app-sandbox</key>
	<true/>
	<key>com.apple.security.cs.allow-jit</key>
	<true/>
	<key>com.apple.security.network.server</key>
    <true/>
    <key>com.apple.security.network.client</key>
    <true/>
</dict>
</plist>' > macos/Runner/DebugProfile.entitlements



# get dependencies
flutter pub get

# initialize wasm and build wasm runtime, this is why we need rust
flutter pub run wasm:setup

# build and run the wasm example
flutter run -d macos
```

**Comment by Hixie** [comment]
If anyone would like to submit a PR to fix the example to add the relevant entitlements, I would be more than happy to review it and land it for you.

**Comment by Kounex** [comment]
I'm unfortunately running into the same issue by trying to run the example/wasm app on macOS (Intel):

```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: WasmError: Wasm library not found. Did you `dart run wasm:setup`?
#0      _getLibPath (package:wasm/src/runtime.dart:119:3)
#1      new WasmRuntime._init (package:wasm/src/runtime.g.dart:118:52)
#2      runtime (package:wasm/src/runtime.dart:20:29)
#3      runtime (package:wasm/src/runtime.dart)
#4      new WasmModule (package:wasm/src/module.dart:21:14)
#5      _ExampleState._loadLogic (package:rfw_wasm/main.dart:62:14)
```

The app itself is starting but the window is empty and therefore the additional content by `rfw` is not loaded. Using `dart run wasm:setup` did not change this behaviour

**Comment by Hixie** [comment]
applying the fix in https://github.com/flutter/packages/pull/2958 should resolve it

**Comment by dkbast** [comment]
Unfortunately, actually it's not - only with this fix applied you get so far that you see this issue. Without it it won't even download the wasm code.

The problem is a) the compiled wasm could not be download without the entitlement and b) it does not recognise that wasm was properly setup- the pr only fixes a)

**Comment by Hixie** [comment]
Ah, interesting. That wasn't clear. Would you mind filing a new issue to track that? I've no idea what could be special about macOS that would make it not work...

**Comment by rodydavis** [comment]
I have been working on this package and wanted to show an alternative approach:

https://github.com/rodydavis/flutter_json_widgets

I am trying to follow the flutter SDK api for creating widgets and being able to create server driven UIs but using dart classes and a runtime parser. (Which can convert to json and have copyWith, map, when methods too)

I think that I could output rfw files, but just found out about this package. Looks really interesting!

**Comment by Hixie** [comment]
> This is exactly what I needed to display banners. Smaller than images but flexible. Could you provide an rfw script converter that converts a Flutter widget to an rfw script language in text or blob?

Like, something that parses Dart code of the form seen in build methods, and outputs rfw files?

**Comment by njovy** [comment]
@Hixie Yes! Dart code to rfw. It can also check some sort of validations with registered widgets. It's hard to figure out how to write a script file for rfw until you read the library code.

**Comment by Hixie** [comment]
Compiling completely arbitrary Dart code into a declarative format is mathematically impossible (via transposition to the halting problem) but one could certainly imagine writing a parser that handles a subset of Dart and a fixed set of Flutter widgets, and turned that into the RFW binary format. Because it would have to be a subset, though, I'm not sure how useful it would be in practice, you'd probably end up spending as much time trying to learn what the subset is as you would just learning to write RFW's dedicated text format directly.

**Comment by linyongsheng** [comment]
@Hixie Whether there is a dedicated IDE plug-in for RFW's text format. Write RFW's text format directly  is low efficiency.

**Comment by Hixie** [comment]
I'm not aware of that currently, but I would be very happy to provide advice to anyone who is interested in writing such a thing!

**Comment by rodydavis** [comment]
> I have been working on this package and wanted to show an alternative approach:
> 
> https://github.com/rodydavis/flutter_json_widgets
> 
> I am trying to follow the flutter SDK api for creating widgets and being able to create server driven UIs but using dart classes and a runtime parser. (Which can convert to json and have copyWith, map, when methods too)
> 
> I think that I could output rfw files, but just found out about this package. Looks really interesting!

Ended up writing a blog post on my explorations into rfw!

https://rodydavis.com/posts/flutter-ssr/

TLDR: Really love it and i think there is a lot of great use cases.

I also think instead of WASM, QuickJS could be a good option!

https://pub.dev/packages/flutter_qjs

**Comment by rodydavis** [comment]
Followed up with logic via [flutter_js](https://pub.dev/packages/flutter_js) on my blog post and example!

https://rodydavis.com/posts/flutter-ssr/

This allows the logic to be shipped on iOS/MacOS via JavaScripCore!

**Comment by maks** [comment]
Nice one @rodydavis ! 👍🏻  There are lots of great packages that could be used here.  For example I'm working with LuaDardo for scripting rfw (which has a few benefits such as being in pure Dart and easy to hack on) and am actually using Lua as a bi-directional  codegen target in my project since its a nice simple syntax.

**Comment by maks** [comment]
@rodydavis  yes thanks again for publishing your example! I think your article is **super** helpful for anyone wanting to get into using RFW 👍🏻 

And yes AppStore review does limit things somewhat, though while its *just* my personal interpretation of it, I believe interpreters such as LuaDardo should be allowed and I believe C impl of Lua is already used in a number of published games.

**Comment by rodydavis** [comment]
Of course! Happy to have made the article because I was trying to understand it too.

I almost made it with this package:

https://pub.dev/packages/hetu_script

**Comment by ralph-bergmann** [comment]
I just discovered this package today, and it sounds like exactly what I was looking for. 

But I saw that it doesn't support theming at all. Is there a technical reason for this? Or do you not have the time to implement it? I need it for the application I have in mind and would spend time implementing it. But I don't want to waste my time. So, if you know it won't work, let me know :-)

**Comment by rodydavis** [comment]
In my example I show how to use theming. You would also have a theme higher than the remote widget with custom themes or provide custom widget resolvers to add more.

**Comment by Hixie** [comment]
It can more or less support any Flutter widget, if you provide a "local widget" for it. Assuming you mean theming in the sens of controlling the theme from the server: I didn't implement the `Theme` widget in https://github.com/flutter/packages/blob/main/packages/rfw/lib/src/flutter/material_widgets.dart mostly because `Theme` has a very large API surface that has been in active flux in recent years, but if anyone wants to submit a PR to add Theme to the widgets implemented by `createMaterialWidgets()` I'm sure we would be happy to review it. You can also implement a subset of it in your own application, it's relatively straightforward and would only have to support the theming aspects you need.

**Comment by johnmccutchan** [comment]
I'd like to be able to reflect over the contents of DynamicContent. My use case is to aid in debugging.

Something like:

```
class DynamicContent {
  void forEach((key, value) => ...);
}
```

would be sufficient. Alternatively, if the root map of DynamicContent were exposed I could just write the code myself.

**Comment by johnmccutchan** [comment]
The object `args` should be explicitly documented. You sort of discover it in the docs but it isn't ever defined. Did I miss something?

**Comment by johnmccutchan** [comment]
Could we get a textual format of DynamicContent? It would be nice for development to be able to paste a bunch of different DynamicContent values into a text field and have it update the root map.

**Comment by Hixie** [comment]
Can you elaborate on your debugging needs?

**Comment by johnmccutchan** [comment]
I'm trying to make a realtime editor and visualizer, left side is a text box where I can type in a rfw text, right side is the rendered widget.

I'd like to also include a display of DynamicContent.

I'm also interested in source maps so I can associate rfw text sections with widgets and DynamicContent values.

**Comment by Hixie** [comment]
If you're building a visualizer, you could just subscribe to the various keys that you're rendering. I guess the problem is that you don't want to hard-code the root keys, so that the visualizer can be used in contexts with different root key conventions? We could let you subscribe with an empty key and have that return the root...

**Comment by Hixie** [comment]
Can you elaborate on what you mean by "source maps" in this context?

**Comment by johnmccutchan** [comment]
> If you're building a visualizer, you could just subscribe to the various keys that you're rendering. I guess the problem is that you don't want to hard-code the root keys, so that the visualizer can be used in contexts with different root key conventions? We could let you subscribe with an empty key and have that return the root...

correct, I want to not have anything hard coded.

**Comment by johnmccutchan** [comment]
> Can you elaborate on what you mean by "source maps" in this context?

imagine I have some rfw text that creates button widget with text fetchec from DynamicContent 
using the key "foo.bar".

I'd like to be able to track what sections of the rfw text map to the Button (or any widget). Also which keys in the DynamicContent are used by the widget.

Think Chrome DevTools where you can hover over a DOM element and navigate to the HTML tag in the source view and vice versa.

**Comment by Hixie** [comment]
Can you file two bugs, one for each of these requests? We can look into what it would take to implement that.

**Comment by tugorez** [comment]
Filed https://github.com/flutter/flutter/issues/141658 to add support for widget builders.

**Comment by tugorez** [comment]
Filed https://github.com/flutter/flutter/issues/141665 to discuss wheter it makes sense to break the existing parser implementation into multiple subpackages to facilitate evolving the RFW language.

**Comment by tugorez** [comment]
Filed https://github.com/flutter/flutter/issues/141666 to discuss RFW ergonomics

**Comment by tugorez** [comment]
Filed https://github.com/flutter/flutter/issues/141668 to register even more local widgets in RFW.

**Comment by tugorez** [comment]
Filed https://github.com/flutter/flutter/issues/141669 to track adding more example apps to RFW.

**Comment by tugorez** [comment]
Filed https://github.com/flutter/flutter/issues/141670 to track migrating some tests to use dart testing utils (instead of flutter testing utils).

**Comment by tugorez** [comment]
Filed https://github.com/flutter/flutter/issues/141709 to notify developers about missmatching arguments.

**Comment by tugorez** [comment]
Filed https://github.com/flutter/flutter/issues/141948 to discuss/explore alternatives to `rfwtxt`.

**Comment by tugorez** [comment]
@Hixie what happens if an RFW library imports two other libraries that contain a widget with the same name? Does the RFW compiler throw? One shadows the other? Do we have a mechanism to deal with that in place? Curious if you gave it a thought already. Also happy to file a new issue to discuss the language itself. 

```
import core1; // Has a container definition
import core2; // Has a container defition as well

widget root = Container();
```

BTW I'm new to github so please let me know if I'm using gh issues the wrong way by using them as subthreads of this discussion.

**Comment by Hixie** [comment]
One shadows the other. There's a TODO somewhere about having enabling prefixes if it becomes a significant problem.

**Comment by ScottS2017** [comment]
Regarding my experience and improving the Dev Exp, this package uses a docs approach that is seen in many places and it unfortunately doesn't always help.

E.G.: I'm trying to work with the core widgets and the syntax for doing so isn't clear. The examples given work fine, but when I start trying to add more parameters and other widgets, things break down. So I naturally go looking through the docs, in particular, createCoreWidgets( ) at [https://pub.dev/documentation/rfw/latest/rfw/createCoreWidgets.html](https://pub.dev/documentation/rfw/latest/rfw/createCoreWidgets.html). There we find a list of widgets and English language descriptions of parameter syntax for exceptions to the norm. Since the English descriptions are not easy to visualize for someone who has never seen this package before, even if the person is a native English speaker like me, I clicked on some links for the widgets with the expectation I would find an RFW JSON example of how to implement each widget I clicked on.

Instead, I found they're just links back to the standard widget catalog. This is not helpful, because anyone working with RFW's core widgets is already going to be more than a little familiar with the listed widgets such as Container, Text, and so on. We don't need a refresher on how the normal widgets work, we need to see how to make each widget work with RFW.

As for the English language descriptions, here is the one for color:
_"[Color](https://api.flutter.dev/flutter/dart-ui/Color-class.html) values are represented as integers. The hex literal values are most convenient for this, the alpha, red, green, and blue channels map to the 32 bit hex integer as 0xAARRGGBB."_

But what we actually need is:
_To use a [Color](https://api.flutter.dev/flutter/dart-ui/Color-class.html) parameter, simply use the hex value of the color without declaring a Color or Colors, like this:  "color: 0xFFFF00FF,"_

What prompted this reply is that I can't seem to make a TextStyle work no matter what I try. There is an English language entry in the docs for createCoreWidgets( ) that states, _"For each, every parameter is implemented using the same name. Parameters that take structured types are represented using maps, with each named parameter of that type's default constructor represented by a key, with the following notable caveats and exceptions..."_

When I read that my gut reaction was, "What on earth does that mean?? I'm more confused now than before I read that..."

TextStyle is not listed as an exception, so apparently I'm supposed to use a map in some way that I can't find an sample of in the examples "Hello" or Local (Remote and WASM don't contain JSON examples in their main.dart files.)

Here is the example of Text from the Hello example:
_"Text(text: ["Hello, ", data.greet.name, "!"], textDirection: "ltr"),"_

So far so good, until I try to use the same approach to change the color of the text:
Text(text: ["Hello, ", data.greet.name, "!"], textDirection: "ltr", style: TextStyle(color: 0xFF00FF00),),

This causes an error not in the Text, but in the widget after the Text: _"Expected symbol ":" but found ( at line..."_

Instead of just trying to find out about Text, I thought it better to go searching for a way to understand this whole approach because I'm sure there are dozens of other situations to which the same issue would apply.

In short, the very first thing a user should see after the general description of any package are examples of exactly which syntax the developer has to type on the keyboard in order to use it that package in an app. We need that far more than English language descriptions, and for people who aren't trained formally as engineers the lack of this in so much documentation is a never ending source of blockers, frustration, and slow downs. I can't even begin to tell you how often I read docs and think, _"Okay, that's what it's doing at a high level but how do I make it **do something** in my code? **What exact syntax do I need to type with my fingers in order to get this thing to work?**_

**Comment by Hixie** [comment]
yeah, sorry for the poor quality documentation. there was a lot to document and i ran out of steam before getting to examples.

**Comment by uberchilly** [comment]
I didn't find a way to show double or int variables passed into the widget with the Text widget, because it only supports String and disregards other types, but there is also no easy way to inline convert double/int to string, or is there? One solution would be to provide local widgets like DoubleText and IntText, and do necessary conversions there, but is there any other way of doing this?

**Comment by ScottS2017** [comment]
@Hixie is there any chance you could just make a one screen widget that includes uses of color, padding (AnimatedPadding, I know), gradient, and other such things? I'm just thinking you could make a "kids watercolor" where you throw in a bunch of junk that doesn't really go together, but just shows us what syntax to use for each thing that's not done the same way normal Flutter code is done. I'm so very close to having Gemini be able to build almost anything. I've got it working already, but I keep hitting blockers that are only because I can't give it examples of what syntax to use. It's able to use anything I've been able to feed it, but there are a lot of things I don't know about this syntax.

I don't know a use case for Gemini controlling the screen remotely, and building it in whatever way the AI decides to, but you have to admit the idea is fascinating. Now I just have to figure out how to take it from parlor trick to something useful.

**Comment by ScottS2017** [comment]
Here is a sample widget showing how some parameters are used. I plan to update this as I keep learning more.

Column(
  children: [
    SizedBox(
      width: 100.0,
      height: 4.0,
      child: ColoredBox(
        color: 0xFF0000FF, // Colors are just the hex value of the color.
      ),
   ),
    Container(
      width: 100.0,
      height: 50.0,
      color: 0xFFFFFF00, // Colors are just the hex value of the color.
      padding: [16.0,10.0,6.0,10.0,], // Padding is just four doubles. The order is LTRB.
      child: Text(
        text: ["Hello World"], // The text parameter is named, and is a list of strings.
        textDirection: "ltr",
        // The style parameter is a map. The keys TextStyle parameters. Note the values like color are still used without the Color or Colors constructors.
        style: {
          'color': 0xFF00FF00, // Colors are just the hex value of the color.
          'fontSize': 24.0,
        },
      ),
    ),
    Icon(
      icon: 0xE2A0, // Icons are just the hex value of the icon, don't use the IconData constructor.
      fontFamily: "MaterialIcons", // Specify the icon font family.
      color: 0xFFFF00FF, // Colors are just the hex value of the color.
      size: 30.0,
    ),
  ],
),

**Comment by uberchilly** [comment]
I want to contribute material widget -> Slider to the library but I don't have Linux machine to create and run the golden tests and they are skipped on mac and windows machines that I have, so my code coverage is not 100% without golden tests. What can I do?

**Comment by Hixie** [comment]
Join the Discord and beg someone to do it for you. :-)
See also https://github.com/flutter/flutter/issues/147187

**Comment by uberchilly** [comment]
You can check my pull request for Slider widget

**Comment by ScottS2017** [comment]
A couple of points:

1) There are use cases that have to utilize rfw text, particularly LLMs that can only generate a text string. Yes, such an implementation is well underway. Please keep the text approach alive and kicking, despite it's inferior performance.

2) The guide I started above is also well underway, and is currently over 750 lines of code. It has to be that way because to get an LLM to work with RFW I have to explain everything about the syntax and how it works in the prompt, and I have to explain it as if the LLM is a three-year-old. Once it's done, or close to, I can make a gist and link it from here.

**Comment by Hixie** [comment]
For use cases that generate text from LLMs, the recommended way to transmit the data to the client is to encode the generated text as the binary format on the server, and transmit that. That said, given the speed of LLMs today, the resulting performance improvement in likely insignificant.

**Comment by arcticfox1919** [comment]
I have a question that I hope can be answered. Is it necessary for all code to be written within a single rfwtxt file? When dealing with complex UI nesting, I would prefer to encapsulate the code into different rfwtxt source files, rather than writing everything in one rfwtxt file.

**Comment by ScottS2017** [comment]
> For use cases that generate text from LLMs, the recommended way to transmit the data to the client is to encode the generated text as the binary format on the server, and transmit that. That said, given the speed of LLMs today, the resulting performance improvement in likely insignificant.

I'm a little confused by this because LLMs can only output text. I haven't been using a server, how would that work?

And I agree, the performance choke point is always going to be the LLM.

**Comment by rodydavis** [comment]
Made a RFW realtime editor!

https://rodydavis.github.io/rfw_studio/

**Comment by Hixie** [comment]
@arcticfox1919 
> I have a question that I hope can be answered. Is it necessary for all code to be written within a single rfwtxt file?

Nope, rfwtxt supports imports.

@ScottS2017
> LLMs can only output text. I haven't been using a server, how would that work?

You would take the rfwtxt output from the LLM, encode it in the binary format, and send that.

**Comment by JeffOlajos** [comment]
We are developers of Flutter Markup Language. and would like to discuss possibilities re: integration of RFW into FML. I can be reached via email at jeff@appdaddy.co or on discord @appdaddy.

**Comment by subzero911** [comment]
Thanks Flutter team, we really need a server-driven UI. At least 1-2 times a week I see such requests in my community.

I would like to have a VSCode extension for rfwtxt syntax highlighting / autocompletion.

**Comment by brsblge** [comment]
Recently, I've been playing around this cool package. I experienced an issue about widget builders and filed #161544.

@tugorez I guess somehow arrays sent from source are not treated as arrays on remote scope.

/cc: @Hixie

**Comment by abdulrehman0796** [comment]
Is their a exhaustive list of all the keyword and widget supported by RFW

**Comment by littleGnAl** [comment]
Is there any plan to generate the `.rfwtxt` file using the existing widget? This would make the `rfw` package developer-friendly without requiring expertise in the `rfw` format. I think the `rfw` format should be a intermediate artifact, and most developers don’t need to know about it.

**Comment by Hixie** [comment]
@littleGnAl Can you elaborate on your proposal?

**Comment by littleGnAl** [comment]
@Hixie 

> The following idea is currently just a brainstorm; whether it is actually feasible still needs to be assessed.

We can use annotations and [analyzer](https://pub.dev/packages/analyzer) to turn existing `Widget`s into `rfw`-format files.

Below are the main annotation definitions:

```dart
/// Annotation to mark a field as a `rfw` data source.
/// The field that marked as `@RfwData` will be generated as `data.<filed-name>`
/// in `rfw` format.
class RfwData {
  const RfwData();
}

/// Annotation to mark a field as a `rfw` args.
/// The field that marked as `@RfwArg` will be generated as `args.<filed-name>`
class RfwArg {
  const RfwArg({this.isEvent = false});
  final bool isEvent;
}

/// Annotation to mark a class as a `rfw` widget builder.
///
/// - The [widgetName] parameter is used to specify the name of the widget that
///   will be generated in `rfw` format. If the [widgetName] is not provided, the
///   class name will be used as the widget name.
///
/// - The [isLocal] parameter is used to specify whether the widget is a local widget.
class RfwBuilder {
  const RfwBuilder({this.widgetName = '', this.isLocal = false});
  final String widgetName;
  final bool isLocal;
}
```

## Basic usage

Mark a `Widget` with `@RfwBuilder`; during generation we extract the code inside its build method and produce the corresponding `rfw` format:

```dart
class MyRfwData {
  const MyRfwData(this.name);
  final String name;
}

@RfwBuilder(widgetName: 'root')
class RootWidget extends StatelessWidget {
  const RootWidget({super.key, required this.greet});

  @RfwData()
  final MyRfwData greet;

  @override
  Widget build(BuildContext context) {
    return Container(
      color: const Color(0xFF002211),
      child: Center(
        // Now we have a type-safe data, `greet.name` will be generated as `data.greet.name`
        // in `rfw` format.
        child: Text('Hello ${greet.name}!', textDirection: TextDirection.ltr),
      ),
    );
  }
}
```

The code above generates the following:

```
// The "import" keyword can be auto added
import core.widgets;
// The name: "root" is specified as `@RfwBuilder(widgetName: 'root')`
widget root = Container(
    color: 0xFF002211,
    child: Center(
    child: Text(text: ["Hello, ", data.greet.name, "!"], textDirection: "ltr"),
    ),
);
```

### Handlers

We use `RfwEvent` to carry extra information:

```dart
/// A class to represent an event in `rfw` format.
/// This allow us to produce `event "<event-name>" { <key-values> }` in `rfw` format.
class RfwEvent {
  const RfwEvent(this.name, this.arguments, this.handler);
  final String name;
  final DynamicMap arguments;
  final VoidCallback handler;
}

@RfwBuilder()
class MyButton extends StatelessWidget {
  const MyButton({super.key, required this.label, required this.onPressed});

  @RfwArg()
  final String label;

  @RfwArg()
  final RfwEvent onPressed;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      // When we found the `handler` is used, we can generate the `args.onPressed` directly.
      // (Actually the `handler` field is useless, it's just used to let use know it's an event handler, 
      // and also let the compilation passed.)
      onTap: onPressed.handler,
      child: Text(label),
    );
  }
}

@RfwBuilder()
class MyEventWidget extends StatelessWidget {
  const MyEventWidget({super.key});

  @override
  Widget build(BuildContext context) {
    return MyButton(
      label: 'Tap me',
      onPressed: RfwEvent('digit', <String, Object>{
        'arguments': [7],
      }, () {}),
    );
  }
}
```

This generates:

```
import core.widgets;

widget MyButton = GestureDetector(
  onTap: args.onPressed,
  child: Text(text: args.label),
);

widget MyEventWidget = MyButton(
    label: "Tap me",
    onPressed: event "digit" { arguments: [7] }
);
```

### Local widgets

To generate a local widget, set `@RfwBuilder(isLocal: true)`:

```dart
@RfwBuilder(isLocal: true)
class GreenBox extends StatelessWidget {
  const GreenBox({super.key, required this.child});

  final Widget child;

  @override
  Widget build(BuildContext context) {
    return ColoredBox(color: const Color(0xFF002211), child: child);
  }
}

@RfwBuilder(isLocal: true)
class Hello extends StatelessWidget {
  const Hello({super.key, required this.name});

  final String name;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Text(
        'Hello, $name!', // source.v<String>(<Object>["name"])
        textDirection: TextDirection.ltr,
      ),
    );
  }
}
```

For local widgets we do not emit an `rfw` format file directly; instead we generate a `LocalWidgetLibrary` map:

```dart
/// These codes are generated.
Map<String, LocalWidgetBuilder> _generatedLocalWidgetsMap() {
  return <String, LocalWidgetBuilder>{
    'GreenBox': (BuildContext context, DataSource source) {
      return GreenBox(child: source.child(<Object>['child']));
      // return ColoredBox(
      //   color: const Color(0xFF002211),
      //   child: source.child(<Object>['child']),
      // );
    },
    'Hello': (BuildContext context, DataSource source) {
      return Hello(name: source.v<String>(<Object>["name"])!);
      // return Center(
      //   child: Text(
      //     'Hello, ${source.v<String>(<Object>["name"])}!',
      //     textDirection: TextDirection.ltr,
      //   ),
      // );
    },
  };
}
```

Add the generated map into a `LocalWidgetLibrary`:

```dart
WidgetLibrary _createLocalWidgets() {
  return LocalWidgetLibrary({
    ..._generatedLocalWidgetsMap(),
    // Hand written local widgets can be added here
  });
}
```

I don’t yet have a clear idea for handling rfw format with State, but I believe it should be feasible to generate rfw format for a `StatefulWidget` using annotations.

**Comment by Hixie** [comment]
Interesting. That seems challenging to do generally (at some level, it reduces to the halting problem), but I imagine it would be doable within some constraints.

There's no current plan to do this that I'm aware of, but if this is something someone is passionate about implementing it, I'd be happy to advise if that's helpful. I'm @Hixie on Discord, feel free to reach out.
